File: src/agent/export.zig
// Copyright The Fantastic Planet - By David Clabaugh
//
// export.zig — Export conversation history in standard formats
//
// Exports conversation history as JSONL for external consumption
// (fine-tuning, analysis, backup).
//
// Formats:
//   history  → JSONL (one message per line from SQLite)
//
// Called from REPL via /export <format> <path>

const std = @import("std");
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const storage_mod = @import("storage.zig");

/// Export conversation history as JSONL.
/// Format: {"role":"user|assistant","content":"...","sequence":N}
pub fn exportHistory(
    alloc: Allocator,
    storage: *storage_mod.Storage,
    path: []const u8,
) !usize {
    // Get recent conversations
    const convos = try storage.listConversations(100);
    defer storage.freeConversations(convos);

    if (convos.len == 0) return 0;

    var buf: ArrayList(u8) = .{};
    const w = buf.writer(alloc);
    var total: usize = 0;

    for (convos) |convo| {
        const messages = storage.loadConversation(convo.id) catch continue;
        defer storage.freeMessages(messages);

        for (messages) |msg| {
            try std.fmt.format(w,
                \\{{"role":"{s}","content":{s},"sequence":{d}}}
            , .{
                msg.role,
                msg.content_json,
                msg.sequence_num,
            });
            try w.writeAll("\n");
            total += 1;
        }
    }

    const jsonl = try buf.toOwnedSlice(alloc);
    defer alloc.free(jsonl);

    const file = try std.fs.cwd().createFile(path, .{});
    defer file.close();
    try file.writeAll(jsonl);

    return total;
}

File: agent-harness/cli_anything/wintermolt/core/export.py
"""Export functionality for Wintermolt conversations.

Exports conversation history to JSONL format, either via the wintermolt
binary or by reading SQLite directly.
"""

import json
from pathlib import Path
from typing import Optional

from cli_anything.wintermolt.core import history


def export_history(
    output_path: str,
    conversation_id: Optional[str] = None,
    limit: int = 100,
) -> dict:
    """Export conversation history to JSONL.

    If conversation_id is given, exports only that conversation.
    Otherwise exports the most recent `limit` conversations.
    """
    output = Path(output_path)
    total = 0

    if conversation_id:
        conversations = [{"id": conversation_id}]
    else:
        conversations = history.list_conversations(limit=limit)

    with output.open("w") as f:
        for convo in conversations:
            messages = history.get_messages(convo["id"])
            for msg in messages:
                record = {
                    "conversation_id": convo["id"],
                    "role": msg.get("role", ""),
                    "content": msg.get("content_json", ""),
                    "sequence": msg.get("sequence_num", 0),
                }
                f.write(json.dumps(record) + "\n")
                total += 1

    return {
        "path": str(output),
        "conversations": len(conversations),
        "messages": total,
    }
